src/palace/manager/service/analytics/eventdata.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import annotations
+
+from datetime import datetime
+from functools import cached_property
+from uuid import UUID
+
+import flask
+from pydantic import (
+    AwareDatetime,
+    BaseModel,
+    ConfigDict,
+    computed_field,
+    field_serializer,
+)
+from pydantic_core.core_schema import FieldSerializationInfo
+from typing_extensions import Self
+
+from palace.manager.sqlalchemy.model.library import Library
+from palace.manager.sqlalchemy.model.licensing import LicensePool
+from palace.manager.sqlalchemy.model.patron import Patron
+from palace.manager.util.datetime_helpers import utc_now
+from palace.manager.util.log import LoggerMixin
+
+
+class AnalyticsEventData(BaseModel, LoggerMixin):
+    """
+    This class represents the data that is sent to the analytics provider.
+
+    It is a Pydantic model that is used to validate the data before it is sent. It stores
+    all the data without references to any database objects, so it can be serialized, stored
+    and sent outside a database transaction.
+    """
+
+    type: str
+    start: AwareDatetime
+
+    # TODO: We include the 'end' field as a copy of 'start' because that is
+    #   what the pre-pydantic implementation did. Since this is just duplicated
+    #   we should remove it in a future release.
+    @computed_field  # type: ignore[prop-decorator]
+    @cached_property
+    def end(self) -> AwareDatetime:
+        return self.start
+
+    library_id: int
+    library_name: str
+    library_short_name: str
+    old_value: int | None
+    new_value: int | None
+
+    @computed_field  # type: ignore[prop-decorator]
+    @cached_property
+    def delta(self) -> int | None:
+        if self.new_value is None or self.old_value is None:
+            return None
+        return self.new_value - self.old_value
+
+    # TODO: This field has been removed from the database model, but is still included
+    #   in the response for backwards compatibility. It should be removed in a future
+    #   release.
+    location: None = None
```

Comment:
@dbernstein Do we need the "location" field in json for redshift? Or can this just be dropped?